libs/core/synchronization/include/hpx/synchronization/counting_semaphore.hpp

```diff
@@ -22,41 +22,379 @@
+    ///        semaphore, since a binary semaphore is better known as a
+    ///        mutex. A counting semaphore is a counter for a set of
+    ///        available resources, rather than a locked/unlocked flag of a
+    ///        single resource. It was invented by Edsger Dijkstra.
+    ///        Semaphores are the classic solution to preventing race
+    ///        conditions in the dining philosophers problem, although they
+    ///        do not prevent resource deadlocks.
+    ///
+    ///        Counting semaphores can be used for synchronizing multiple
+    ///        threads as well: one thread waiting for several other
+    ///        threads to touch (signal) the semaphore, or several threads
+    ///        waiting for one other thread to touch this semaphore. Unlike
+    ///        \a hpx::mutex a \a counting_semaphore_var is not tied to threads
+    ///        of execution -- acquiring a semaphore can occur on a
+    ///        different thread than releasing the semaphore, for example.
+    ///        All operations on \a counting_semaphore_var can be performed
+    ///        concurrently and without any relation to specific threads of
+    ///        execution, with the exception of the destructor which cannot
+    ///        be performed concurrently but can be performed on a different
+    ///        thread.
+    ///
+    ///        Semaphores are lightweight synchronization primitives used
+    ///        to constrain concurrent access to a shared resource. They
+    ///        are widely used to implement other synchronization
+    ///        primitives and, whenever both are applicable, can be more
+    ///        efficient than condition variables.
+    ///
+    ///        A counting semaphore is a semaphore object that models a
+    ///        non-negative resource count.
+    ///
+    ///        Class template \a counting_semaphore_var maintains an internal
+    ///        counter that is initialized when the semaphore is created.
+    ///        The counter is decremented when a thread acquires the
+    ///        semaphore, and is incremented when a thread releases the
+    ///        semaphore. If a thread tries to acquire the semaphore when
+    ///        the counter is zero, the thread will block until another
+    ///        thread increments the counter by releasing the semaphore.
+    ///
+    ///        Specializations of \a hpx::counting_semaphore_var are not
+    ///        \namedrequirement{DefaultConstructible},
+    ///        \namedrequirement{CopyConstructible},
+    ///        \namedrequirement{MoveConstructible},
+    ///        \namedrequirement{CopyAssignable},
+    ///        or \namedrequirement{MoveAssignable}.
+    ///
+    /// \note \a counting_semaphore_var's \a try_acquire() can spuriously fail.
+    ///
+    /// \tparam Mutex Type of mutex
+    /// \tparam N     The initial value of the internal semaphore lock
+    ///               count.
+    ///
```

Comment:
`binary_semaphore` is now in its own file (as discussed). I think you should move the corresponding documentations there as well.